### PR TITLE
Update destination service to send NoEndpoints updates

### LINKED ISSUE #563: The Destination service should notify the proxy when the service doesn't exist
See #155. The proxy needs to know when the service doesn't exist (independently of whether it has any endpoints available) so that it can decide whether to fall back to DNS. A PR to document the syntax and semantics of the gRPC API change between the proxy and the destinations service will be provided shortly.

### PR DESCRIPTION
**Pull request by klingerf**
In this branch I'm updating the destination service to start sending NoEndpoints events when a client subscribes, and when a service no longer exists.

The destination service code isn't heavily tested, but I manually verified this change using the [destination-client script](https://github.com/runconduit/conduit/blob/kl/no-endpoints/controller/script/destination-client/main.go).

If I subscribe before a service exists, and then create the service, I see the following events:

```
$ ./bin/go-run controller/script/destination-client -path books.default.svc.cluster.local:7002
INFO[0000] NoEndpoints:                                 
INFO[0000] - exists:false                               
INFO[0000]                                              
INFO[0023] Add:                                         
INFO[0023] - 172.17.0.13:7002                           
INFO[0023]                                              
INFO[0025] Add:                                         
INFO[0025] - 172.17.0.12:7002                           
INFO[0025]                                              
INFO[0025] Add:                                         
INFO[0025] - 172.17.0.10:7002                           
```

When I then delete the service, I see:

```
INFO[0044] Remove:                                      
INFO[0044] - 172.17.0.10:7002                           
INFO[0044]                                              
INFO[0044] Remove:                                      
INFO[0044] - 172.17.0.12:7002                           
INFO[0044] - 172.17.0.13:7002                           
INFO[0044]                                              
INFO[0046] NoEndpoints:                                 
INFO[0046] - exists:false                               
```

---

Whereas if I subscribe after the service has already been created, then I see the following events:

```
$ ./bin/go-run controller/script/destination-client -path books.default.svc.cluster.local:7002
INFO[0000] NoEndpoints:                                 
INFO[0000] - exists:true                                
INFO[0000]                                              
INFO[0000] Add:                                         
INFO[0000] - 172.17.0.10:7002                           
INFO[0000] - 172.17.0.12:7002                           
INFO[0000] - 172.17.0.13:7002                           
```

When I then delete the service, I see:

```
INFO[0009] Remove:                                      
INFO[0009] - 172.17.0.13:7002                           
INFO[0009]                                              
INFO[0009] Remove:                                      
INFO[0009] - 172.17.0.10:7002                           
INFO[0009] - 172.17.0.12:7002                           
INFO[0009]                                              
INFO[0010] NoEndpoints:                                 
INFO[0010] - exists:false                               
```

Fixes #563.

### CODE REVIEW

File: controller/gen/public/api.pb.go [hunk 1]
```diff
@@ -206,9 +206,7 @@ func (m *MetricValue) String() string            { return proto.CompactTextStrin
 func (*MetricValue) ProtoMessage()               {}
 func (*MetricValue) Descriptor() ([]byte, []int) { return fileDescriptor0, []int{2} }
 
-type isMetricValue_Value interface {
```
Comment (siggy) [inline controller/gen/public/api.pb.go]: a lot of these pb changes are just whitespace. just want to confirm this is expected / an artifact of a different protoc version?

Comment (klingerf) [reply controller/gen/public/api.pb.go]: Hmm, yeah, this is a good point. As far as I can tell, the previously generated bindings weren't using the latest protoc-gen-go lib. If I run `protoc-go.sh` on master, I see the same changes as the ones in this branch:

```
kl ~/workspace/conduit (master) $ ./bin/protoc-go.sh && git diff --stat
 controller/gen/common/common.pb.go                  | 24 ++++++------------------
 controller/gen/controller/telemetry/telemetry.pb.go | 13 +++++++++----
 controller/gen/proxy/destination/destination.pb.go  |  4 +---
 controller/gen/proxy/tap/tap.pb.go                  | 16 ++++------------
 controller/gen/proxy/telemetry/telemetry.pb.go      |  4 +---
 controller/gen/public/api.pb.go                     |  8 ++------
 6 files changed, 23 insertions(+), 46 deletions(-)
```

What happens if you run that script on master?

Comment (siggy) [reply controller/gen/public/api.pb.go]: yup, this isn't specific to your branch. some of these files have been in this state for months, also our versioning of `github.com/golang/protobuf/protoc-gen-go` in `Gopkg.lock` hasn't changed either. i suppose we should merge it.

```bash
(master)$ git diff --stat
 controller/gen/common/common.pb.go                  | 24 ++++++------------------
 controller/gen/controller/telemetry/telemetry.pb.go | 13 +++++++++----
 controller/gen/proxy/destination/destination.pb.go  |  4 +---
 controller/gen/proxy/tap/tap.pb.go                  | 16 ++++------------
 controller/gen/proxy/telemetry/telemetry.pb.go      |  4 +---
 controller/gen/public/api.pb.go                     |  8 ++------
 6 files changed, 23 insertions(+), 46 deletions(-)
```

**Comment by adleong** [comment]
I think this makes sense.  Just out of curiosity, does this mean that every stream will always start with a NoEndpoints message?  Is it necessary to send the initial NoEndpoints message if the service does exist?  Is there a semantic difference between `NoEndpoints{exists=true}` and the absence of any messages?  As far as I understand, in both of those cases the client should not fall back to DNS.

**Comment by klingerf** [comment]
@adleong I should have linked to it in the PR description, but there's a longer description of this API in #564, here: https://github.com/runconduit/conduit/pull/564/files#diff-f2176355288276a6d05ceb390603f5e9R13

**Comment by adleong** [comment]
Thanks!  According to that description, `NoEndpoints(true); add` is equivalent to `add` so I'm not sure I understand why we send the `NoEndpoints(true)` at all.

**Comment by briansmith** [comment]
> Thanks! According to that description, NoEndpoints(true); add is equivalent to add so I'm not sure I understand why we send the NoEndpoints(true) at all.

I created the original design specifically to allow `NoEndpoints(exists)` to be sent as the first message by the Destination server, regardless of the value of `exists`, because I thought that simplifies the implementation of the Destination service. However, based on the discussion in #573, I'm going to revise the protocol so that we can't take that shortcut by disallowing `NoEndpoints(true)` when there are >0 endpoints.

File: controller/k8s/endpoints.go [hunk 2]
```diff
@@ -131,7 +133,7 @@ func (e *EndpointsWatcher) GetService(service string) (*v1.Service, bool, error)
 // Watches a Kubernetes resource type
 type informer struct {
 	informer cache.Controller
-	store    *cache.Store
+	store    cache.Store
```
Comment (klingerf) [inline controller/k8s/endpoints.go]: This field was a pointer to an interface, which meant that we had to do some funky dereferencing to access it (e.g. `(*e.serviceInformer.store).GetByKey`). Have switched it to just an interface, no pointer.

File: controller/k8s/endpoints.go [hunk 3]
```diff
@@ -395,43 +374,38 @@ func (sp *servicePort) updateService(newService *v1.Service) {
...
 		sp.addresses = newAddresses
 	}
 }
 
-func (sp *servicePort) deleteService() {
-	sp.mutex.Lock()
-	defer sp.mutex.Unlock()
-
-	newTargetPort := intstr.FromInt(int(sp.port))
-	if newTargetPort != sp.targetPort {
```
Comment (klingerf) [inline controller/k8s/endpoints.go]: As far as I can tell, the `deleteService` method was a no-op. The delete event returns the previous version of the service, so this if condition always returns false. It's also the case that when we delete a service, we receive both a DeleteEndpoints event and a DeleteService event, so I've removed the DeleteService handler, in favor of only having a DeleteEndpoints handler.

**Pull request merged**